ESCPOS_NET/Emitters/BaseCommandEmitter/ImageCommands.cs

```diff
@@ -58,100 +55,35 @@ public byte[] BufferImage(byte[] image, int maxWidth = -1, bool isLegacy = false
-                    byte widthH = (byte)(byteWidth >> 8);
-                    imageCommand.Append(new byte[] { Cmd.GS, Images.ImageCmdLegacy, 0x30, 0x00, widthL, widthH, heightL, heightH });
-                }
-                else
-                {
-                    byte widthL = (byte)(width);
-                    byte widthH = (byte)(width >> 8);
-                    imageCommand.Append(new byte[] { 0x30, 0x70, 0x30, 0x01, 0x01, colorByte, widthL, widthH, heightL, heightH });
-                }
-
-
-                // TODO: test making a List<byte> if it's faster than using ByteArrayBuilder.
-                double[] ditherErrorsNext = new double[width + 2];
-                // Bit pack every 8 horizontal bits into a single byte.
-                for (int y = 0; y < img.Height; y++)
-                {
-                    // Create an array for capturing dither errors
-                    double[] ditherErrorsCurrent = ditherErrorsNext;
-                    ditherErrorsNext = new double[width + 2];
+                imageData = img.ToSingleBitPixelByteArray(maxWidth: maxWidth == -1 ? (int?)null : maxWidth);
+                height = img.Height;
+                width = img.Width;
+            }
 
-                    Span<Rgba32> pixelRowSpan = img.GetPixelRowSpan(y);
-                    byte buffer = 0x00;
-                    int bufferCount = 7;
-                    for (int x = 0; x < img.Width; x++)
-                    {
-                        // Determine if pixel should be colored in.
-                        var pixelValue = ((0.30 * pixelRowSpan[x].R) + (0.59 * pixelRowSpan[x].G) + (0.11 * pixelRowSpan[x].B) + (ditherErrorsCurrent[x]));
-                        double error = 0;
-                        if (pixelValue <= 127)
-                        {
-                            // Pixel value should be black (0), if it is greater than zero then the error is the difference (pixelvalue - 0)
-                            // which is just the pixel value.
-                            error = pixelValue;
-                            buffer |= (byte)(0x01 << bufferCount);
-                        }
-                        else
-                        {
-                            // Pixel should be white, so error is pixelvalue - 255
-                            error = pixelValue - 255;
-                        }
+            byte heightL = (byte)height;
+            byte heightH = (byte)(height >> 8);
 
-                        // Propagate dither.
-                        //error <<= 1; // error * 32
-                        double pixelValue8 = error / 4;
-                        double pixelValue4 = error / 8;
-                        double pixelValue2 = error / 16;
-                        ditherErrorsCurrent[x + 1] += pixelValue8;
-                        ditherErrorsCurrent[x + 2] += pixelValue4;
+            if (isLegacy)
+            {
+                var byteWidth = (width + 7 & -8) / 8;
```

Comment:
could you explain what this calculation is doing with the bitwise and of a negative number?

Reply:
That's the most efficient way I found to round next greater multiple of 8. https://books.google.es/books?id=iBNKMspIlqEC&lpg=PP1&hl=es&pg=PA45#v=onepage&q&f=false

Do you prefer to use `while (n % 8 != 0) n++` which is more readable?